homeassistant/components/matrix.py

```diff
@@ -0,0 +1,352 @@
+"""
+The matrix bot component.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/matrix/
+"""
+import logging
+import os
+from functools import partial
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.notify import (ATTR_TARGET, ATTR_MESSAGE)
+from homeassistant.const import (CONF_USERNAME, CONF_PASSWORD,
+                                 CONF_VERIFY_SSL, CONF_NAME)
+from homeassistant.util.json import load_json, save_json
+
+REQUIREMENTS = ['matrix-client==0.1.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+SESSION_FILE = 'matrix.conf'
+
+CONF_HOMESERVER = 'homeserver'
+CONF_ROOMS = 'rooms'
+CONF_COMMANDS = 'commands'
+CONF_WORD = 'word'
+CONF_EXPRESSION = 'expression'
+
+EVENT_MATRIX_COMMAND = 'matrix_command'
+
+DOMAIN = 'matrix'
+
+COMMAND_SCHEMA = vol.All(
+    # Basic Schema
+    vol.Schema({
+        vol.Exclusive(CONF_WORD, 'trigger'): cv.string,
+        vol.Exclusive(CONF_EXPRESSION, 'trigger'): cv.is_regex,
+        vol.Required(CONF_NAME): cv.string,
+        vol.Optional(CONF_ROOMS, default=[]): vol.All(cv.ensure_list,
+                                                      [cv.string]),
+    }),
+    # Make sure it's either a word or an expression command
+    cv.has_at_least_one_key(CONF_WORD, CONF_EXPRESSION)
+)
+
+PLATFORM_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOMESERVER): cv.url,
+    vol.Optional(CONF_VERIFY_SSL, default=True): cv.boolean,
+    vol.Required(CONF_USERNAME): cv.matches_regex("@[^:]*:.*"),
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_ROOMS, default=[]): vol.All(cv.ensure_list,
+                                                  [cv.string]),
+    vol.Optional(CONF_COMMANDS, default=[]): [COMMAND_SCHEMA]
+})
+
+SERVICE_SEND_MESSAGE = 'send_message'
+
+SERVICE_SCHEMA_SEND_MESSAGE = vol.Schema({
+    vol.Required(ATTR_MESSAGE): cv.string,
+    vol.Required(ATTR_TARGET): vol.All(cv.ensure_list, [cv.string]),
+})
+
+
+def setup(hass, config):
+    """Set up the Matrix bot component."""
+    if not config[DOMAIN]:
+        return False
+
+    config = config[DOMAIN][0]
+
+    bot = MatrixBot(
+        hass,
+        os.path.join(hass.config.path(), SESSION_FILE),
+        config.get(CONF_HOMESERVER),
+        config.get(CONF_VERIFY_SSL),
+        config.get(CONF_USERNAME),
+        config.get(CONF_PASSWORD),
+        config.get(CONF_ROOMS, []),
+        config.get(CONF_COMMANDS, []))
+    hass.data[DOMAIN] = bot
```

Comment:
Possible to verify here that we were able to connect and that username/pass are valid?

Reply:
I thought that components shouldn't do I/O in their constructor, so I moved all the login stuff into the `_setup` method, which is run as a scheduled job. Should I move login back to the constructor? That way, we could indeed verify that we can log in at this point…

Reply:
It's fine to do I/O during the setup phase.